subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -128,18 +133,20 @@ public UnitOfWork getWork() {
             });
     }
 
-    private static void cleanupTreesAfterLoadFailure(UnitOfWork work) {
-        work.visitOutputTrees((name, type, root) -> {
-            try {
-                if (root.exists()) {
-                    if (root.isDirectory()) {
-                        FileUtils.cleanDirectory(root);
-                    } else {
-                        FileUtils.forceDelete(root);
+    private static void cleanOutputsAfterLoadFailure(UnitOfWork work) {
+        work.visitOutputProperties((name, type, roots) -> {
+            for (File root : roots) {
+                try {
+                    if (root.exists()) {
+                        if (root.isDirectory()) {
+                            FileUtils.cleanDirectory(root);
```

Comment:
I think we have a cleaner now which has a better error message when it fails to clean up some file. Should we use this maybe?

Reply:
What new cleaner do you have in mind?

Reply:
This one: https://github.com/gradle/gradle/blob/21925d22551636a6578cc343118a41e91bc9bb75/subprojects/core/src/main/java/org/gradle/api/internal/file/delete/Deleter.java

Reply:
Would you mind opening an issue about this?

Reply:
See https://github.com/gradle/gradle/issues/8829.